ArjunaCore/arjuna/classes/com/arjuna/ats/internal/arjuna/recovery/PeriodicRecovery.java

```diff
@@ -265,6 +231,40 @@ public Mode suspendScan (boolean async)
    {
        synchronized (_stateLock)
        {
+           // see if suspension should delay until certain types have recovered
+           if (recoveryPropertyManager.getRecoveryEnvironmentBean().isWaitForFinalRecovery()) {
+               // List<String> types = recoveryPropertyManager.getRecoveryEnvironmentBean().getTypeNamesToBlockShutdown();
+
+               // disable the transaction system but keep recovery running
+               TxControl.disable();
+               // wait for in flight transactions to complete
+               TransactionReaper.terminate(true);
+
```

Comment:
Looking at this more, this doesn't feel like the right place to control the transaction manager side of things.

Did you see this class before: https://github.com/jbosstm/narayana/blob/main/ArjunaJTS/integration/src/main/java/com/arjuna/ats/jbossatx/jta/TransactionManagerService.java (there is a JTS version to look at too). You can see it getting stopped (I don't know if that is helpful to know) from WildFly here (at least hopefully, unless the lines move around after I got the link): https://github.com/wildfly/wildfly/blob/main/transactions/src/main/java/org/jboss/as/txn/service/ArjunaTransactionManagerService.java#L137

Reply:
> Did you see this class before

Yeah, I think in one of the iterations of this PoC we discussed `TransactionManagerService`. I even modify it in one of my implementations. For example: [1](https://github.com/jmfinelli/narayana/commit/53125ab5caf41b4ad5a774031f4bfd90fe8454a5) and [2](https://github.com/jmfinelli/narayana/commit/8ea6429a575bf38a657ec22db9a86a144e6deb43).

> Looking at this more, this doesn't feel like the right place to control the transaction manager side of things.

Yeah, if this is what you're referring to, I think that exposing this behaviour to integrating parties is the right choice, as Jonathan [suggested](https://github.com/jmfinelli/personal-narayana-proposals/pull/1#issuecomment-1240702395): `Lifecycle management belongs in the server, not the transaction engine. Recovery needs to expose sufficient state/callbacks for the server to determine when its done, but driving it is the server's problem.`